tiledb/sm/misc/resource_pool.h

```diff
@@ -0,0 +1,121 @@
+/**
+ * @file   resource_pool.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2020-2021 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This defines a pool of resources that can be shared for example amongst
+ * threads.
+ */
+
+#ifndef RESOURCE_POOL_H
+#define RESOURCE_POOL_H
+
+namespace tiledb {
+namespace sm {
+
+/**
+ * @tparam T The resource type
+ * @tparam P Class P<T> provides instances of ThreadResource<T>
+ */
+template <class T, template <class> class P>
+class Resource {
+ public:
+  /** Constructor. */
+  Resource(P<T>& p, unsigned int n)
+      : p_(p)
+      , n_(n) {
+  }
+
+  /** Destructor. */
+  ~Resource() {
+    p_.release(n_);
+  }
+
+  /** Get the resource. */
+  T& get() {
+    return p_.at(n_);
+  }
+
+ private:
+  /** The pool that issued this handle. */
+  P<T>& p_;
+
+  /** The index of a vector within the pool to the resource. */
+  unsigned int n_;
+};
+
+template <class T>
+class ResourcePool {
+ public:
+  /** Constructor. */
+  ResourcePool(unsigned int n) {
+    resources_.resize(n);
+    unused_.reserve(n);
+    for (unsigned int i = 0; i < n; i++)
+      unused_.push_back(i);
```

Comment:
No need for `push_back` which changes the size each time. Can use `unused[i]=i` if the size is already set.